### PR TITLE
Create CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by s-perron**
This is intended to be a document that people who want to help the project can get a quick overview of how things are done.  Please comment and suggest any changes.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	git remote add <userid> <url>
+	git fetch <userid> <branch>
+	git checkout <userid>/<branch>
+	git branch -D <branch> # if it already exists
+	git checkout -b <branch>  # assuming it does not collide.
+    ```
+  
+*   Rebase the branch on top of the master branch.  If there are conflicts, send it back to the author and ask them to rebase.  During the interactive rebase be sure to squash all of the commits down to a single commit.
+
+	  ```
```
Comment (antiagainst) [inline CONTRIBUTING.md]: Indent?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    ```
+    git@github.com:<userid>/SPIRV-Tools.git
+    ```
+
+
+*   \<branch\> is the name of the branch that contains the PR
+*   It is assumed that upstream points to [git@github.com](mailto:git@github.com):KhronosGroup/SPIRV-Tools.git.  You can check this by doing
+
+    	`git remote -v`
```
Comment (antiagainst) [inline CONTRIBUTING.md]: Use triple backticks here? Single ones are generally for inline stuff.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+*   Make sure your local copy of the master branch is up to date:
+
+    ```
+    git pull
+    ```
+
+*   Get a local copy of the branch containing the PR.
+  
+    ```
+	git remote add <userid> <url>
```
Comment (antiagainst) [inline CONTRIBUTING.md]: Indent?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+To report a new bug or request a new feature, please file a GitHub issue. Please ensure the bug has not already been reported by searching [issues](https://github.com/KhronosGroup/SPIRV-Tools/issues) and [projects](https://github.com/KhronosGroup/SPIRV-Tools/projects). If the bug has not already been reported open a new one [here](https://github.com/KhronosGroup/SPIRV-Tools/issues/new).
+
+
+When opening a new issue for a bug, make sure you provide the following:
+
+
+
+*   A clear and descriptive title.
+    *   We want a title that will make it easy for people to remember what the issue is about.  Simply using "Segfault in spir-opt" is not helpful because there could be (but hopefully aren't) multiple bugs with segmentation faults with different causes.
```
Comment (antiagainst) [inline CONTRIBUTING.md]: s/spir-opt/spirv-opt/

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+*   A description of the problem that needs to be solved.
+*   Examples that demonstrate the problem.
+
+
+## For developers: Contributing a patch
+
+Before we can use your code, you must sign the [Khronos Open Source Contributor License Agreement](https://cla-assistant.io/KhronosGroup/SPIRV-Tools) (CLA), which you can do online. The CLA is necessary mainly because you own the copyright to your changes, even after your contribution becomes part of our codebase, so we need your permission to use and distribute your code. We also need to be sure of various other things -- for instance that you'll tell us if you know that your code infringes on other people's patents. You don't have to sign the CLA until after you've submitted your code for review and a member has approved it, but you must do it before we can put your code into our codebase.
+
+See README.md for instruction on how to get, build, and test the source.  Once you have made your changes:
```
Comment (antiagainst) [inline CONTRIBUTING.md]: Link for README.md?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+*   Changes are made to a container while iterating through it.  You have to be careful that iterators are not invalidated or that elements are not skipped.
+*   C++11 and VS2013.  We generally assume that we have a C++11 compliant compiler.  However, on Windows, we still support Visual Studio 2013, which is not fully C++11 compliant. See [here](https://msdn.microsoft.com/en-us/library/hh567368.aspx).  In particular, note that it does not provide default move-constructors or move-assignments for classes.  In general, r-value references do not work the way you might assume they do.
+*   For SPIR-V transforms: The module is changed, but the analyses are not updated.  For example, a new instruction is added, but the def-use manager is not updated.  Later on, it is possible that the def-use manager will be used, and give wrong results.
+
+
+## For maintainers: Merging a PR
+
+We intend to maintain a linear history on the GitHub master branch, and the build and its tests should pass at each commit in that history.  A linear always-working history is easier understand and it easier to bisect in case we want to find which commit introduced a bug.
```
Comment (dneto0) [inline CONTRIBUTING.md]: "and it easier to bisect" --> "and easier to bisect"

File: CONTRIBUTING.md (hunk 4 above)
Comment (s-perron) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 5 above)
Comment (s-perron) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 6 above)
Comment (s-perron) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 2 above)
Comment (s-perron) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 3 above)
Comment (s-perron) [reply CONTRIBUTING.md]: fixed.

File: CONTRIBUTING.md (hunk 1 above)
Comment (s-perron) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+*   Create a pull request (PR) with your patch.
+*   Make sure the PR description clearly identified the problem, explains the solution, and references the issue if applicable.
+*   Watch the continuous builds to make sure they pass.
+*   Request a code review.
+
+The reviewer can either approve your PR or request changes.  If changes are requested:
+
+
+
```
Comment (dnovillo) [inline CONTRIBUTING.md]: Add:

* Please make sure your code follows the Google C++ Style Guide.
* For help formatting your code, you can run the command 'clang-format -style=file -i [modified-files]'

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    git branch -D <branch> # if it already exists
+    git checkout -b <branch>  # assuming it does not collide.
+    ```
+  
+*   Rebase the branch on top of the master branch.  If there are conflicts, send it back to the author and ask them to rebase.  During the interactive rebase be sure to squash all of the commits down to a single commit.
+
+    ```
+    git branch -u master
+    git rebase -i
+    ```
```
Comment (dnovillo) [inline CONTRIBUTING.md]: I intend to contribute my script that does all these steps.  I'll add it as an alternative later, or we could simply point folks to it.

File: CONTRIBUTING.md (hunk 7 above)
Comment (s-perron) [reply CONTRIBUTING.md]: Fixed.

File: CONTRIBUTING.md (hunk 8 above)
Comment (s-perron) [reply CONTRIBUTING.md]: We can create a reference to your script when it if available.  For now, I'll leave it as is.

**Comment by s-perron** [comment]
Ready for another review.  I don't think the build errors are related to this change.  How could they be?

**Comment by dnovillo** [comment]
On Mon, Nov 27, 2017 at 10:29 PM, Steven Perron <notifications@github.com>
wrote:

> Ready for another review. I don't think the build errors are related to
> this change. How could they be?
>

They aren't.  Windows builds have been broken for a while due to a missing
#include <cstdint> in source/val/decoration.h (fixed now by
https://github.com/KhronosGroup/SPIRV-Tools/pull/1018) .

File: CONTRIBUTING.md (hunk 8 above)
Comment (dnovillo) [reply CONTRIBUTING.md]: SGTM

**Comment by s-perron** [comment]
Rebase, squashed, and pushed as d78e11c.

**Pull request closed without merging**